File: src/main/java/PijActor.java
import org.agilewiki.jactor.*;
import org.agilewiki.jactor.lpc.JLPCActor;
import org.agilewiki.jactor.lpc.Request;

import java.math.BigDecimal;
import java.math.BigInteger;
import java.util.Random;

/**
 * @author Daniel Seidler
 * @since 2013/01/04
 */
public class PijActor implements PiCalculator{
    public void calculate(final int nrOfWorkers, final int nrOfElements, final int nrOfMessages) throws Exception {
        System.out.println("JACTOR - calc - ("+nrOfWorkers+","+nrOfElements+","+nrOfMessages+")");
        MailboxFactory mailboxFactory = JAMailboxFactory.newMailboxFactory(nrOfWorkers);
        Mailbox mailbox = mailboxFactory.createMailbox();
        Master master = new Master();
        master.initialize(mailbox);
        JAFuture future = new JAFuture();
        Start.init(nrOfElements,nrOfMessages).send(future,master);
        mailboxFactory.close();
    }
}

class Start extends Request<Object, Master> {
    private int nrOfElements;
    private int nrOfMessages;

    private Start(int nrOfElements, int nrOfMessages){
       this.nrOfElements = nrOfElements;
        this.nrOfMessages = nrOfMessages;
    }

    public static Start init(int nrOfElements, int nrOfMessages){
      return new Start(nrOfElements, nrOfMessages);
    }

    public void processRequest(JLPCActor targetActor, final RP rp)
            throws Exception {
        Master a = (Master) targetActor;
        RP<BigDecimal> prp = new RP<BigDecimal>() {
            final long start = System.currentTimeMillis();
            boolean pending = true;
            BigDecimal pi = BigDecimal.ZERO;
            int nrOfResults = 0;

            public void processResponse(BigDecimal result) throws Exception {
                pi = pi.add(result);
                nrOfResults += 1;
                if (nrOfResults == nrOfMessages) {
                    long duration = System.currentTimeMillis() - start;
                    System.out.println(String.format("\tPi approximation: \t\t%s\n\tCalculation time: \t%s milliseconds", pi, duration));
                    rp.processResponse(null);
                }
            }
        };
        a.calculate(nrOfMessages,nrOfElements, prp);
    }

    public boolean isTargetType(Actor targetActor) {
        return targetActor instanceof Master;
    }
}


class Master extends JLPCActor {
    public void calculate(final int nrOfMessages, int nrOfElements, final RP rp) throws Exception {
        for (int start = 0; start < nrOfMessages; start++) {
            Worker worker = new Worker();
            worker.initialize(getMailboxFactory().createAsyncMailbox());
            (new Work(start, nrOfElements)).send(this, worker, rp);
        }
    }
}


class Worker extends JLPCActor {
    public BigDecimal calculatePiFor(int start, int nrOfElements) {
        BigDecimal acc = BigDecimal.ZERO;
        for (int i = start * nrOfElements; i <= ((start + 1) * nrOfElements - 1); i++) {
            acc = acc.add(new BigDecimal(4.0 * (1 - (i % 2) * 2) / (2 * i + 1)));
        }
        return acc;
    }
}


class Work extends Request<BigDecimal, Worker> {
    private int start;
    private int nrOfElements;

    public Work(int start, int nrOfElements) {
        this.start = start;
        this.nrOfElements = nrOfElements;
    }

    @Override
    public boolean isTargetType(Actor actor) {
        return actor instanceof Worker;

    }

    @Override
    public void processRequest(JLPCActor targetActor, RP rp) throws Exception {
        Worker worker = (Worker) targetActor;
        BigDecimal result = ((Worker) targetActor).calculatePiFor(start, nrOfElements);
        rp.processResponse(result);
    }
}

File: src/main/java/kotlin/actor/PiJActor.kt
package kotlin.actor

import java.math.BigDecimal
import org.agilewiki.jactor.lpc.Request
import org.agilewiki.jactor.lpc.JLPCActor
import org.agilewiki.jactor.RP
import org.agilewiki.jactor.JAMailboxFactory
import org.agilewiki.jactor.JAFuture
import akka.dispatch.BoundedMailbox
import org.agilewiki.jactor.Mailbox

/**
 * Created with IntelliJ IDEA.
 * User: daniel
 * Date: 06.01.13
 * Time: 18:00
 * To change this template use File | Settings | File Templates.
 */

fun main(args: Array<String>) {
    PijActor().calculate(5,1000,1000)
}
public class PijActor: PiCalculator{
    public override fun calculate(nrOfWorkers : Int, nrOfElements: Int, nrOfMessages: Int){
        System.out.println("JACTOR - calc - (" + nrOfWorkers + "," + nrOfElements + "," + nrOfMessages + ")");
        val mailboxFactory = JAMailboxFactory.newMailboxFactory(nrOfWorkers!!)!!;
        val mailbox = mailboxFactory.createMailbox()!!
        val future = JAFuture();
        MasterA(nrOfElements, nrOfMessages, mailbox).send(future);
        mailboxFactory.close();
    }
}

class MasterA(val nrOfElements: Int, val nrOfMessages: Int,mailbox: Mailbox) : Actor<BigDecimal, Any>(mailbox){
    protected override fun processRequest(rp: RP<BigDecimal>) {
        val prp = object: RP<BigDecimal>() {
            val start = System.currentTimeMillis();
            var pi = BigDecimal.ZERO
            var  nrOfResults = 0;
            public override fun processResponse(result: BigDecimal?) {
                pi = pi.add(result!!);
                nrOfResults += 1;
                if (nrOfResults == nrOfMessages) {
                    val duration = System.currentTimeMillis() - start;
                    System.out.println(java.lang.String.format("\tPi approximation: \t\t%s\n\tCalculation time: \t%s milliseconds", pi, duration));
                    (rp as RP<Any>).processResponse(null);
                }
            }
        };
        0..(nrOfMessages-1) forEach {
            val mailbox = getMailboxFactory()!!.createAsyncMailbox()!!;
            WorkerA(it,nrOfElements, mailbox).send(this,prp)
        }
    }
}

class WorkerA(val start: Int, val nrOfElements: Int,mailbox: Mailbox) : Actor<BigDecimal, BigDecimal>(mailbox){
    protected override fun processRequest(rp: RP<BigDecimal>) {
        var acc = BigDecimal.ZERO
        ((start * nrOfElements)..((start + 1) * nrOfElements - 1)) forEach {
            acc = acc.add(BigDecimal(4.0 * (1 - (it % 2) * 2) / (2 * it + 1)))
        }
        rp.processResponse(acc)
    }
}